File: WdtcCore/src/main/java/org/wdt/wdtc/core/launch/GameCLICommand.java
package org.wdt.wdtc.core.launch;


import com.google.gson.JsonElement;
import org.wdt.utils.io.FileUtils;
import org.wdt.wdtc.core.auth.User;
import org.wdt.wdtc.core.game.GameVersionJsonObject;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.game.config.DefaultGameConfig;
import org.wdt.wdtc.core.manger.VMManger;

import java.io.IOException;
import java.util.Map;

public class GameCLICommand extends AbstractGameCommand {
  private final Launcher launcher;

  public GameCLICommand(Launcher launcher) {
    this.launcher = launcher;
  }

  public StringBuilder getCommand() throws IOException {
    DefaultGameConfig.Config gameConfig = launcher.getGameConfig().getConfig();
    GameVersionJsonObject VersionJsonObject = launcher.getGameVersionJsonObject();
    nonBreakingSpace(VersionJsonObject.getMainClass());
    for (JsonElement Element : VersionJsonObject.getArguments().getGameList()) {
      if (!Element.isJsonObject()) {
        nonBreakingSpace(ReplaceData(Element.getAsString()));
      }
    }
    nonBreakingSpace("--height");
    nonBreakingSpace(gameConfig.getHight());
    nonBreakingSpace("--width");
    command.append(gameConfig.getWidth());
    return command;
  }

  private Map<String, String> getDataMap() throws IOException {
    User user = User.getUser();
    return Map.of("${auth_player_name}", user.getUserName(), "${version_name}", launcher.getVersionNumber(),
        "${game_directory}", FileUtils.getCanonicalPath(launcher.getVersionPath()), "${assets_root}", FileUtils.getCanonicalPath(launcher.getGameAssetsDirectory()),
        "${assets_index_name}", launcher.getGameVersionJsonObject().getAssets(), "${auth_uuid}", user.getUuid(),
        "${auth_access_token}", user.getAccessToken(), "${user_type}", user.getType().toString(),
        "${version_type}", "Wdtc-" + VMManger.getLauncherVersion());
  }


  private String ReplaceData(String str) throws IOException {
    Map<String, String> ReplaceMap = getDataMap();
    for (String s : ReplaceMap.keySet()) {
      str = str.replace(s, ReplaceMap.get(s));
    }
    return str;
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/launch/GameCLICommand.kt
package org.wdt.wdtc.core.launch

import org.wdt.wdtc.core.auth.preferredUser
import org.wdt.wdtc.core.game.Launcher
import org.wdt.wdtc.core.game.config.gameConfig
import org.wdt.wdtc.core.manger.launcherVersion
import java.io.IOException

class GameCLICommand(private val launcher: Launcher) : AbstractGameCommand() {
  @Throws(IOException::class)
  override fun getCommand(): StringBuilder {
    val gameConfig = launcher.gameConfig.config!!
    val versionJsonObject = launcher.gameVersionJsonObject
    nonBreakingSpace(versionJsonObject.mainClass!!)
    for (element in versionJsonObject.arguments?.gameList!!) {
      if (!element.isJsonObject) {
        nonBreakingSpace(replaceData(element.asString))
      }
    }
    nonBreakingSpace("--height")
    nonBreakingSpace(gameConfig.hight)
    nonBreakingSpace("--width")
    commandBuilder.append(gameConfig.width)
    return commandBuilder
  }

  private val dataMap
    get() = mutableMapOf(
      "\${auth_player_name}" to
          preferredUser.userName,
      "\${version_name}" to
          launcher.versionNumber,
      "\${game_directory}" to
          launcher.versionDirectory.canonicalPath,
      "\${assets_root}" to
          launcher.gameAssetsDirectory.canonicalPath,
      "\${assets_index_name}" to
          launcher.gameVersionJsonObject.assets,
      "\${auth_uuid}" to
          preferredUser.uuid,
      "\${auth_access_token}" to
          preferredUser.accessToken,
      "\${user_type}" to
          preferredUser.type.toString(),
      "\${version_type}" to
          "Wdtc-$launcherVersion"
    )


  @Throws(IOException::class)
  private fun replaceData(strs: String): String {
    var str = strs
    for (s in dataMap.keys) {
      str = str.replace(s, dataMap[s]!!)
    }
    return str
  }
}

